### PR TITLE
feat(Core/Packets): Port packet handling from TrinityCore

### LINKED ISSUE #3670: Feature: Adding IP address in the log of malformed packets [$15]
Feature: Adding IP address in the log of malformed packets

Yesterday something happened in the server was all normal and from one moment to another began to generate lag throughout the world, expelled many people went to check the Server.Log and came out this message:

image

You can see that it does not give any information, is there any possibility to add the IP address in that LOG?
in the mangos page i think they talked about it

https://www.getmangos.eu/forums/topic/3828-worldsockethandle_input_header-client-sent-malformed-packet-size/

They give some information but the structure is different.

image

We would appreciate it very much if you could detect the IP through the log.

<bountysource-plugin>

---
There is a **[$15 open bounty](https://www.bountysource.com/issues/94282650-feature-adding-ip-address-in-the-log-of-malformed-packets?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #5575: Core/Packet: Port TC Packet handling
It would be nice to port https://github.com/WarheadCore/WarheadBand/commit/847df113f1e0ed9ead2213704731f6efca7929aa to avoid conflicts with the "new" stuff we are porting from TC...

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/98247151-core-packet-port-tc-packet-handling?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Port packet handling from TrinityCore

## Issues Addressed:
- Closes #5575
- Closes #3670
- Part #3451

## Tests Performed:
- Test in game

## How to Test the Changes:
- Check packet system
- Do the usual things

## Target Branch(es):
- [x] Master

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Kitzunu** [comment]
I need this to continue https://github.com/azerothcore/azerothcore-wotlk/pull/5428

**Comment by Winfidonarleyan** [comment]
Added #3670

**Comment by Winfidonarleyan** [comment]
Ready to test

**Comment by Sedatyf** [comment]
> Check packet system

What do you mean by "check packet system"? Is there a specific thing to do?

**Comment by Winfidonarleyan** [comment]
> > Check packet system
> 
> What do you mean by "check packet system"? Is there a specific thing to do?

no specific, try bg/instance or other, need check packet sending

**Comment by Sedatyf** [comment]
Hey @Winfidonarleyan 
I played nearly one hour on a server with this PR applied and I had no problem. My server didn't crash and I experienced nothing extraordinary. Maybe 2 or 3 lags/freezes but the source could be my hardware. I did some dungeons and a BG. Unfortunately, I couldn't test it with other friends, just me on two different accounts. 
I don't know if you had other things in mind, if yes, feel free to poke me and I'll do more test on it

**Comment by FrancescoBorzi** [comment]
> Maybe 2 or 3 lags/freezes but the source could be my hardware.

@Sedatyf have you ever experienced the same issue while testing on `master` or other PRs ?

**Comment by Kitzunu** [comment]
How are we looking on this? I still need it merged so I can continue on trainer rewrite

**Comment by Winfidonarleyan** [comment]
Today fix conflicts and go merge

**Comment by Winfidonarleyan** [comment]
Done

File: src/common/Utilities/MessageBuffer.h [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    size_type GetBufferSize() const { return _storage.size(); }
+
+    // Discards inactive data
+    void Normalize()
+    {
+        if (_rpos)
+        {
+            if (_rpos != _wpos)
+                memmove(GetBasePointer(), GetReadPointer(), GetActiveSize());
```
Comment (FrancescoBorzi) [inline src/common/Utilities/MessageBuffer.h:67]: so many missing brackets :)

Comment (Winfidonarleyan) [reply src/common/Utilities/MessageBuffer.h:67]: aah, okay

File: src/common/Utilities/MessageBuffer.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    void ReadCompleted(size_type bytes) { _rpos += bytes; }
+
+    void WriteCompleted(size_type bytes) { _wpos += bytes; }
+
+    size_type GetActiveSize() const { return _wpos - _rpos; }
+
+    size_type GetRemainingSpace() const { return _storage.size() - _wpos; }
+
+    size_type GetBufferSize() const { return _storage.size(); }
```
Comment (FrancescoBorzi) [inline src/common/Utilities/MessageBuffer.h:59]: please apply `[[nodiscard]]` when possible

Comment (Winfidonarleyan) [reply src/common/Utilities/MessageBuffer.h:59]: okay

**Comment by Winfidonarleyan** [comment]
Ready

**Review by Kitzunu — APPROVED** [review]

**Pull request merged**